testng-core/src/main/java/org/testng/internal/ConfigurationGroupMethods.java

```diff
@@ -121,10 +73,40 @@ public List<ITestNGMethod> getBeforeGroupMethodsForGroup(String[] groups) {
     }
   }
 
-  public List<ITestNGMethod> getAfterGroupMethodsForGroup(String group) {
+  public List<ITestNGMethod> getAfterGroupMethods(ITestNGMethod testMethod) {
+    if (testMethod.hasMoreInvocation() || testMethod.getGroups().length == 0) {
+      return Collections.emptyList();
+    }
+
+    Set<String> methodGroups = new HashSet<>(Arrays.asList(testMethod.getGroups()));
+
     synchronized (afterGroupsThatHaveAlreadyRun) {
-      return retrieve(afterGroupsThatHaveAlreadyRun, m_afterGroupsMethods, group);
+      if (m_afterGroupsMap == null) {
+        m_afterGroupsMap = initializeAfterGroupsMap();
+      }
+
+      return methodGroups.stream()
+          .filter(t -> isLastMethodForGroup(t, testMethod))
+          .map(t -> retrieve(afterGroupsThatHaveAlreadyRun, m_afterGroupsMethods, t))
+          .filter(Objects::nonNull)
+          .flatMap(Collection::stream)
+          .filter(t -> isAfterGroupAllowedToRunAfterTestMethod(t, methodGroups))
+          .collect(Collectors.toList());
+    }
+  }
+
+  private boolean isAfterGroupAllowedToRunAfterTestMethod(
```

Comment:
It looks the method has no dependency and it could be static

Reply:
it depends on m_afterGroupsMap